com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationArchiveExtractor.java

```diff
@@ -0,0 +1,200 @@
+        for (ZipEntry entry; (entry = inputStream.getNextEntry()) != null;) {
+            if (!entry.getName()
+                .startsWith(streamProducer.getStreamEntryName())) {
+                continue;
+            }
+            validateEntry(entry.getName());
+            Path filePath = resolveTempEntryPath(entry.getName(), streamProducer.getStreamEntryName(), dirPath);
+            saveEntry(streamProducer.getInputStream(), entry, filePath, callback);
+        }
+    }
+
+    private void copy(InputStream input, OutputStream output, ExtractStatusCallback callback) throws IOException, SLException {
+        byte[] buffer = new byte[BUFFER_SIZE];
+        int numberOfReadBytes = 0;
+        while ((numberOfReadBytes = input.read(buffer)) != -1) {
+            callback.onBytesToWrite(numberOfReadBytes);
+            output.write(buffer, 0, numberOfReadBytes);
+        }
+    }
+
+    private void saveEntry(InputStream inputStream, ZipEntry entry, Path filePath, ExtractStatusCallback callback)
+        throws IOException, SLException {
+        if (!isFile(entry.getName())) {
+            Files.createDirectories(filePath);
+            return;
+        }
+
+        Files.createDirectories(filePath.getParent());
+        filePath = Files.createFile(filePath);
+        try (OutputStream outputStream = Files.newOutputStream(filePath)) {
+            copy(inputStream, outputStream, callback);
+        }
+    }
+
+    private void saveStreamToDirectory(InputStream inputStream, String entryName, String dirName, Path dirPath,
+        ExtractStatusCallback callback) throws IOException, SLException {
+        validateEntry(entryName);
+        Path filePath = resolveTempEntryPath(entryName, dirName, dirPath);
+        Files.createDirectories(filePath.getParent());
+        Files.createFile(filePath);
+        try (OutputStream outputStream = Files.newOutputStream(filePath)) {
+            copy(inputStream, outputStream, callback);
+        }
+    }
+
+    private boolean isFile(String fileName) {
+        return !fileName.endsWith(ARCHIVE_FILE_SEPARATOR);
+    }
+
+    private String getPrefix(String nameOfFile) {
+        int indexOfDot = nameOfFile.lastIndexOf('.');
+        String prefix = (indexOfDot > 0) ? nameOfFile.substring(0, indexOfDot) : nameOfFile;
+        if (prefix.length() < 3) {
+            prefix = prefix + "000";
+        }
+        return prefix;
+    }
+
+    private String getSuffix(String nameOfFile) {
+        int indexOfDot = nameOfFile.lastIndexOf('.');
+        return (indexOfDot > 0) ? nameOfFile.substring(indexOfDot) : null;
+    }
+
+    public static Path getTempDirectory(String entryName) throws IOException {
+        String dirName = entryName;
+        if (dirName.endsWith(ARCHIVE_FILE_SEPARATOR)) {
+            dirName = dirName.substring(0, dirName.length() - 1);
+        }
+        dirName = dirName.replace('/', '-');
+        Path dirPath = Paths.get(dirName);
+        if (Files.exists(dirPath)) {
```

Comment:
Why check whether it exists? And BTW, wouldn't Files.createTempDirectory create a directory with a random string appended to the prefix that is the "dirName" variable. If yes, then this check doesn't make sense, because you're checking whether directory A exists, but if it doesn't you're creating directory B.